### ISSUE TITLE
token mananger generating multiple default tokens per service account

### ISSUE THREAD

**Issue by macb** [kubernetes/kubernetes#24809]
Over the course of multiple clusters of varying age we're seeing additional tokens being generated for service accounts.

Cluster A ~3mo old cluster:
```
$ kubectl get serviceaccount --all-namespaces
NAMESPACE     NAME      SECRETS   AGE
default       default   10        96d
kube-system   default   11        80d

$ kubectl get secrets --all-namespaces
NAMESPACE     NAME                  TYPE                                  DATA      AGE
default       default-token-4uvk8   kubernetes.io/service-account-token   3         60d
default       default-token-7e9ds   kubernetes.io/service-account-token   3         43d
default       default-token-d18wp   kubernetes.io/service-account-token   3         50d
default       default-token-dstki   kubernetes.io/service-account-token   3         33d
default       default-token-dtc7e   kubernetes.io/service-account-token   3         50d
default       default-token-mf0de   kubernetes.io/service-account-token   3         43d
default       default-token-q0p2q   kubernetes.io/service-account-token   3         52d
default       default-token-rsh1a   kubernetes.io/service-account-token   3         40d
default       default-token-s1q8h   kubernetes.io/service-account-token   3         51d
default       default-token-wcxpw   kubernetes.io/service-account-token   3         39d
kube-system   default-token-3v664   kubernetes.io/service-account-token   3         51d
kube-system   default-token-4koh5   kubernetes.io/service-account-token   3         36d
kube-system   default-token-6lezi   kubernetes.io/service-account-token   3         44d
kube-system   default-token-8455x   kubernetes.io/service-account-token   3         39d
kube-system   default-token-8nxi2   kubernetes.io/service-account-token   3         42d
kube-system   default-token-9v8oe   kubernetes.io/service-account-token   3         44d
kube-system   default-token-f5bpn   kubernetes.io/service-account-token   3         36d
kube-system   default-token-hzwb4   kubernetes.io/service-account-token   3         60d
kube-system   default-token-jrdjs   kubernetes.io/service-account-token   3         53d
kube-system   default-token-thqnw   kubernetes.io/service-account-token   3         49d
kube-system   default-token-w9uar   kubernetes.io/service-account-token   3         37d
```

Cluster B ~2mo old:
```
$ kubectl get serviceaccount --all-namespaces
NAMESPACE     NAME      SECRETS   AGE
default       default   1         66d
kube-system   default   5         66d

$ kubectl get secrets --all-namespaces
NAMESPACE     NAME                  TYPE                                  DATA      AGE
default       default-token-5d0gn   kubernetes.io/service-account-token   3         60d
kube-system   default-token-0jpip   kubernetes.io/service-account-token   3         60d
kube-system   default-token-1ra2j   kubernetes.io/service-account-token   3         52d
kube-system   default-token-eemer   kubernetes.io/service-account-token   3         43d
kube-system   default-token-fdjve   kubernetes.io/service-account-token   3         44d
kube-system   default-token-nfpl8   kubernetes.io/service-account-token   3         38d
```

Cluster C ~1mo old:
```
$ kubectl get serviceaccount --all-namespaces
NAMESPACE     NAME      SECRETS   AGE
default       default   1         39d
kube-system   default   1         39d
```

In this case, all secrets are the token controller generated tokens (we aren't using secrets at the moment). It seems like this is possibly causing issues with applications that talk to the API with the given token. One such application (prometheus) is getting back a 403 error after being restarted at some point in the night however skydns seems fine with a ~51d old token in cluster A.

It's also curious that the token generation seems to stop around the time cluster C was introduced. What would cause a new token to be generated? Maybe we did that without realizing it.

Unsure what additional information would be of use, happy to provide whatever would help though I'm not sure we have logs from 39+ days ago.

**Comment by mikedanese** [comment]
@liggitt

**Comment by liggitt** [comment]
what versions are each cluster running? a couple relevant PR's that merged ~2 months ago:
https://github.com/kubernetes/kubernetes/pull/21706
https://github.com/kubernetes/kubernetes/pull/22160

**Comment by liggitt** [comment]
new tokens being generated shouldn't cause 403 errors from old tokens

**Comment by macb** [comment]
Ya the 403's were unrelated and just lead us to find this issue, sorry about that. :)

Cluster A and B are on 1.2.0. Cluster C is on 1.2.1 (but was previously 1.2.0 and upgraded to 1.2.1).

**Comment by liggitt** [comment]
yeah, based on age, it looks like the spurious tokens were created pre-1.2.0, and that definitely matches the symptoms those two pulls addressed. I'm going to call this fixed in 1.2.0, though please ping if you see it again

**Comment by macb** [comment]
Sounds good, thanks for taking a look.
